numpy/random/_generator.pyx

```diff
@@ -4428,6 +4433,14 @@ cdef class Generator:
                     x[i] = buf
         else:
             # Untyped path.
+            if not isinstance(x, MutableSequence):
+                # See gh-18206. We may decide to deprecate here in the future.
+                warnings.warn(
+                    "`x` isn't a recognized object; `shuffle` is not guaranteed "
+                    "to behave correctly. E.g., non-numpy array/tensor objects "
+                    "with view semantics may contain duplicates after shuffling."
+                )
```

Comment:
Maybe needs `stacklevel=2` for a clear context message

Reply:
good catch, will update